### DISCUSSION TITLE
How to deal with WP_Error?

### DISCUSSION THREAD

**Discussion by kilianso** [timber/timber#2601] [Q&A] [answered]
Hey folks

I'm struggling a bit catching the `WP_Error` object in Timber. I have an object oriented custom theme and let logged in users change data with `update_user_meta ` on the `author.php` page. Since i think i'm missing something on the Timber-side, i try to find an answer here. Any hint would be much appreciated.

Here's my `UserMeta.php` Class. I'm loading it with autoload in `Theme.php`.

```php
<?php

namespace SBX\Theme\Package;

class UserMeta
{
   // Instantiate the WP_Error object
   public $error = '';
	
   public function __construct() {
        $this->$error = new \WP_Error();
   }

   public function run() {
        // Update user meta from profile/user/author pages
	add_action( 'template_redirect', [$this, 'updateUserMetaFromFrontend'] );
   }

   // function to update meta fields from auhtor/profile pages
   public function updateUserMetaFromFrontend() {
	
        // Check if the user is authenticated.
        if ( ! is_user_logged_in() ) {
	        return;
        }
        
        // Check if we have all necessary data.
        if ( empty( $_POST['update_meta_nonce'] ) ) {
	        return;
        }
        
        // Verify the nonce.
        if ( ! wp_verify_nonce( $_POST['update_meta_nonce'], 'update-meta' ) ) {
	        return;
        }

        // Update optional user meta
        $this->updateField('nickname', true, 'Please enter a name.');
        $this->updateField('city');
        
        // Redirect back to the previous page.
        wp_safe_redirect( wp_get_referer() );
        exit;
   }

    public function updateField($field, $mandatory = false, $errormsg = null) {
        if (!empty($_POST[$field]) && is_string($_POST[$field])) {
                update_user_meta( get_current_user_id(), $field, htmlspecialchars($_POST[$field]) );
        }else {
            if($mandatory) {
               $this->$error->add('general', '<strong>' . __('ERROR') . '</strong>: ' . $errormsg, array('form-field' => $field));
            }
        }
    }
}

```

My `author.php`

```php
<?php

$context = Timber::context();
$context['errors'] = is_wp_error($error);
Timber::render( array( 'author.twig'), $context );
```

And my `author.twig`

```twig
{% if errors %}
   <pre>
      {{ dump(errors) }}
   </pre>
{% endif %}
<form method="post" action="{{ site.url }}" id="profile-edit-form">
    <fieldset class="ProfileEdit__accounts">
        <label>
            <span>Name<small class="ProfileEdit__required-note">required</small></span>
            <input 
                type="text" 
                name="nickname" 
                placeholder="Enter your name" 
                value="{{user.nickname}}" 
                autocomplete="off" 
                required>
        </label>
        <label>
            <span>City</span>
            <input 
                type="text" 
                name="city" 
                placeholder="Your City" 
                value="{{user.account_city}}" 
                autocomplete="off">
        </label>
    </fieldset>
    {% do function('wp_nonce_field', 'update-meta', 'update_meta_nonce') %}
    <button type="submit" class="submit">Update profile</button>
</form>
```

So the `dump(errors)` in my Twig template is always `false` but i just don't get it why. I thought the `WP_Error` object is accessible within the same route. What am i missing?

**Comment by gchtr** [comment, marked as answer]
I see a couple of things. I think the following line in **author.php** might be the first issue:

```php
$context['errors'] = is_wp_error($error);
```

I don’t see where the `$error` variable is coming from. Did you only publish a part of the template file, or is that it? 

The `errors` variable in Twig is always false, because `$error` is always undefined and `is_wp_error()` returns only true if `$error` is an instance of `WP_Error`.

Also, I don’t see where you actually instantiate the `UserMeta` class. The template file doesn’t automatically have access to the `UserMeta` class. You would have to instantiate it there. I’m thinking something like this might work:

**author.php**

```php
<?php

$user_meta = new UserMeta();

$context = Timber::context();
$context['errors'] = is_wp_error( $user_meta->get_error() );

Timber::render( array( 'author.twig'), $context );
```

But then, your `template_redirect` action won’t work anymore, because that function already ran. But you could also directly call the `updateUserMetaFromFrontend()` method from your template file and remove the `template_redirect` action.

You then have to make sure that the form loads the same page instead of redirecting to the home page. Check the `action` attribute in your `form` tag.

```
<form method="post" action="{{ user.link }}" id="profile-edit-form">
```

You could maybe use a `get_errors()` method in the `UserMeta` class instead of using `is_wp_error()`.

```php
<?php

$user_meta = new UserMeta();

$user_meta->updateUserMetaFromFrontend();

$context = Timber::context();
$context['errors'] = $user_meta->get_errors();

Timber::render( array( 'author.twig'), $context );
```

To iterate further, you could also add `$user_meta` itself to the context:

```php
<?php

$user_meta = new UserMeta();

$user_meta->updateUserMetaFromFrontend();

$context = Timber::context();
$context['user_meta'] = $user_meta;

Timber::render( array( 'author.twig'), $context );
```

In Twig, you could then use:

```twig
{% if user_meta.get_errors is not empty %}
   <pre>
      {{ dump(user_meta.get_errors) }}
   </pre>
{% endif %}
```

Then, the next problem is when you create an empty `WP_Error` object in the `UserMeta` constructor:

```php
public function __construct() {
    $this->$error = new \WP_Error();
}
```

The `$error` property will always return true when you use `is_wp_error()`. You should only create a new `WP_Error` object if there actually is an error. I wouldn’t actually use a constructor at all.

I hope that helps you to get this working ☺️.

**Comment by kilianso** [comment]
Hey @gchtr! Thanks for the detailed answer. Checking for errors with `get_errors()` did the trick. Regarding the whole instantiating of the `UserMeta` Class: This already happens thanks to an autoloader mechanism in my setup which is based on the [helloroots](https://github.com/SayHelloGmbH/hello-roots) starter.

Thanks and Gruess us Bärn 👋

**Reply by gchtr** [reply]
Great, I’m glad it helped!

Happy weekend vo Winterthur nach Bern 👋
